crates/storage-macros/src/lib.rs

```diff
@@ -0,0 +1,290 @@
+                None
+            } else {
+                let getter = format_ident!("{}", ident);
+                let setter = format_ident!("set_{}", ident);
+
+                Some(quote! {
+                    pub fn #getter(&self) -> &#ty {
+                        &self.#ident
+                    }
+
+                    pub fn #setter(&mut self, value: #ty) -> bool {
+                        if self.#ident == value {
+                            false
+                        } else {
+                            self.#ident = value;
+                            self.#storage_ident.update();
+                            true
+                        }
+                    }
+                })
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let serializable_fields: Vec<_> = named_fields
+        .iter()
+        .filter(|f| {
+            !f.attrs
+                .iter()
+                .any(|attr| attr.path().is_ident("skip") || attr.path().is_ident("private"))
+                && f.ident.as_ref().unwrap() != storage_ident
+        })
+        .map(|f| f.ident.as_ref().unwrap())
+        .collect();
+
+    let skipped_fields: Vec<_> = named_fields
+        .iter()
+        .filter(|f| {
+            f.attrs
+                .iter()
+                .any(|attr| attr.path().is_ident("skip") || attr.path().is_ident("private"))
+        })
+        .map(|f| f.ident.as_ref().unwrap())
+        .collect();
+
+    let deserialize_impl = quote! {
+        impl borsh::BorshDeserialize for #name {
+            fn deserialize_reader<R: std::io::Read>(reader: &mut R) -> std::io::Result<Self> {
+                let #storage_ident = #storage_ty::deserialize_reader(reader)?;
+                Ok(Self {
+                    #storage_ident,
+                    #(#serializable_fields: borsh::BorshDeserialize::deserialize_reader(reader)?,)*
+                    #(#skipped_fields: Default::default(),)*
+                })
+            }
+        }
+    };
+
+    let serialize_impl = quote! {
+        impl borsh::BorshSerialize for #name {
+            fn serialize<W: std::io::Write>(&self, writer: &mut W) -> std::io::Result<()> {
+                borsh::BorshSerialize::serialize(&self.#storage_ident, writer)?;
+                #(borsh::BorshSerialize::serialize(&self.#serializable_fields, writer)?;)*
+                Ok(())
+            }
+        }
+    };
+
+    let expanded = quote! {
+        impl #name {
+            #(#field_implementations)*
+        }
+
+        impl calimero_storage::entities::Data for #name {
+            type Child = #child_type;
+
+            fn element(&self) -> &calimero_storage::entities::Element {
+                &self.#storage_ident
+            }
+
+            fn element_mut(&mut self) -> &mut calimero_storage::entities::Element {
+                &mut self.#storage_ident
+            }
+        }
+
+        impl calimero_storage::entities::AtomicUnit for #name {}
```

Comment:
_... This reads like that's the idea. `calimero_storage` won't compile to WASM, nor should we want it to.

But these macro definitions would be used by applications and be compiled to WASM, so you need to bridge between the guest and the host which manages storage.

Reply:
I guess that can be refined later as I don't really know anything about it 🙂

Reply:
That's one of the things I raised for us to clarify while designing the storage, what's needed to connect the guest to the host are storage entry identifiers, and you mentioned we'd have those so it seemed like a solved problem, now we have them in the form of `Path`s.

So all that's left is to expose `calimero_storage` to the `calimero_runtime`, and allow the guest address storage using `Path`s

Reply:
> That sounds like something to look at later down the line...

Well, not quite.. Maybe you can explain how you think of it, perhaps that's where the disconnect is. But the only purpose of this work is to redefine a foundation for the guest state that is easily reconcilable under the asynchronous, optimistically executed model, which affords us the transition from the transactional model.

So without these pieces - exposure to the guest & sync, we can't use it any other way.

Reply:
So to confirm, exposure to the guest is not part of these changes, and nor is the network/context side.